Assets/MixedRealityToolkit/Providers/Hands/HandRay/AltHandRay.cs

```diff
@@ -0,0 +1,164 @@
+        public void Update(Vector3 handPosition, Vector3 palmNormal, Transform headTransform, Handedness sourceHandedness)
+        {
+            Vector3 newStableHeadPose = UpdateStableHeadPose(headTransform);
+            MixedRealityPose stabilizedHeadPose = new MixedRealityPose(newStableHeadPose, headTransform.rotation);
+
+            // Approximate new torso direction.
+            Vector3 newTorsoDirection = ComputeApproximateTorsoDirection(stabilizedHeadPose, handPosition, sourceHandedness);
+
+            // Mix the new torso direction with the last to ease into direction changes.
+            lastTorsoDirection = Vector3.Lerp(newTorsoDirection, lastTorsoDirection, 0.5f);
+
+            Quaternion torsoRotation = Quaternion.LookRotation(lastTorsoDirection, Vector3.up);
+
+            Vector3 rayPivotPoint = ComputeApproximateShoulderPosition(stabilizedHeadPose, torsoRotation, sourceHandedness);
+
+            Vector3 handRayDirection = (handPosition - rayPivotPoint).normalized;
+
+            this.palmNormal = palmNormal;
+            this.headForward = headTransform.forward;
+            this.headUp = headTransform.up;
+
+            stabilizedRay.AddSample(new Ray(handPosition, handRayDirection));
+        }
+
+        private Vector3 UpdateStableHeadPose(Transform headTransform)
+        {
+            // We check the distance from the last received head pose, and keep the previous value if the distance hasn't sufficiently changed.
+            // This is done to prevent shifts to the head position caused by slight head rotation.
+            float distanceFromLastPose = Vector3.Distance(headTransform.position, lastStableHeadPose);
+
+            if (distanceFromLastPose > MinHeadPoseDistanceForCorrection)
+            {
+                float headDistanceOffset = Mathf.Clamp01(distanceFromLastPose / cHeadDistanceThreshold);
+                float correctionAmt = Mathf.Pow(headDistanceOffset, 4f);
+                lastStableHeadPose = Vector3.Lerp(lastStableHeadPose, headTransform.position, correctionAmt);
+            }
+            return lastStableHeadPose;
+        }
+
+        /// <summary>
+        /// Computes an approximate shoulder position based on handedness and head position.
+        /// </summary>
+        private Vector3 ComputeApproximateShoulderPosition(MixedRealityPose headPose, Quaternion torsoRotation, Handedness sourceHandedness)
+        {
+            // Left/Right shoulder
+            float shoulderDirection = (sourceHandedness == Handedness.Left ? -1f : 1f);
+
+            Vector3 shoulderPosition = headPose.Position -
+                                       // Vertical offset
+                                       Vector3.up * ShoulderHeightOffset +
+                                       // Shoulder horizontal offset
+                                       (torsoRotation * Vector3.right * ShoulderWidthOffset * shoulderDirection);
```

Comment:
another minor suggestion

(torsoRotation * Vector3.right * (ShoulderWidthOffset * shoulderDirection));
without the parenthesis, I believe it will be 
quaternion * vector => vector
vector * scalar => vector
vector * scalar => vector

vs the parenthesis merges scalar * scalar saves a few multiplications. 

Again minor, but easy to add
https://docs.unity3d.com/Manual/BestPracticeUnderstandingPerformanceInUnity7.html

Reply:
Interesting! Thanks for the suggestion